Document/0x04g-Testing-Cryptography.md

```diff
@@ -158,8 +160,10 @@ Note: AES-CBC with PKCS #5 has shown to be vulnerable to padding oracle attacks
 When memory dumping is part of your threat model, then keys can be accessed the moment they are actively used. Memory dumping either requires root-access (e.g. a rooted device or jailbroken device) or it requires a patched application with Frida (so you can use tools like Fridump).
 Therefore it is best to consider the following, if keys are still needed at the device:
 
-- make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use Android Keystore) or Secure Enclave (e.g. use the Keychain and when you sign, use ECDHE).
-- If keys are necessary which are outside of the TEE / SE, make sure you obfuscate/encrypt them and only de-obfuscate them during use. Always zero out keys before the memory is released, whether using native code or not. This means: overwrite the memory structure (e.g. nullify the array) and know that most of the Immutable types in Android (such as `BigInteger` and `String`) stay in the heap.
+- Make sure that all cryptographic actions and the keys itself remain in the Trusted Execution Environment (e.g. use Android Keystore) or Secure Enclave (e.g. use the Keychain and when you sign, use ECDHE).
+- If keys are necessary which are outside of the TEE / SE, make sure keys are protected using multi-layered approach. For example, data encryption key (DEK) can be encrypted with key encryption key (KEK) stored in Keychain / Keystore, thus DEK is stored encrypted. This key wrapping technique requires reading KEK first and then decrypting DEK before usage. Another approach is obfuscation: obfuscate DEK and de-obfuscate before usage.
```

Comment:
I suggest we do both!

First, we explain the DEK-KEK approach, because it is platform- and implementation-independent. Anyone can do it.

Second, I also added a link to EncryptedSharedPreferences from `androidx.security.crypto` package as an illustration of this approach. I've seen many similar libs for iOS and Android (encrypted shared preferences, encrypted user defaults), but I'd avoid linking community-written libs (because who knows how they have implemented encryption piece, how they store keys, if they provide any expiration/cleanup, etc :) ). 

Linking `androidx.security.crypto` makes sense because it's made by Google, more trust, more support.

I rewrote this piece, please take a look.

Reply:
Random idea: we can also link to this [Cryptographic Storage Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Cryptographic_Storage_Cheat_Sheet.html#encrypting-stored-keys) that also explains key wrapping technique.

Reply:
Also, Amazon calls this technique "envelope encryption": https://docs.aws.amazon.com/wellarchitected/latest/financial-services-industry-lens/use-envelope-encryption-with-customer-master-keys.html